### PR TITLE
GH-43598: [C++][Parquet] Parquet Metadata Printer supports print sort-columns

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

Now we have "sort-columns" support in Parquet spec, Python ( https://github.com/apache/arrow/pull/37665/files ) and C++. We can support print it in metadata Printer

### What changes are included in this PR?

Add "SortingColumns" support in parquet printer

### Are these changes tested?

* [x] TODO after https://github.com/apache/parquet-testing/pull/56 is merged

### Are there any user-facing changes?

No

* GitHub Issue: #43598

### CODE REVIEW

**Comment by mapleFU** [comment]
After this patch:

```json
{
  "Version": "2.6",
  "CreatedBy": "parquet-cpp-arrow version 16.1.0",
  "TotalRows": "3",
  "NumberOfRowGroups": "1",
  "NumberOfRealColumns": "2",
  "NumberOfColumns": "2",
  "Columns": [
     { "Id": "0", "Name": "a", "PhysicalType": "INT64", "ConvertedType": "NONE", "LogicalType": {"Type": "None"} },
     { "Id": "1", "Name": "b", "PhysicalType": "BYTE_ARRAY", "ConvertedType": "UTF8", "LogicalType": {"Type": "String"} }
  ],
  "RowGroups": [
     {
       "Id": "0",  "TotalBytes": "166",  "TotalCompressedBytes": "174",  "SortColumns": [{"column_idx":0, "descending":1, "nulls_first": 1}, {"column_idx":1, "descending":0, "nulls_first": 0}],  "Rows": "3",
       "ColumnChunks": [
          {"Id": "0", "Values": "3", "StatsSet": "True", "Stats": {"NumNulls": "1", "Max": "2", "Min": "1" },
           "Compression": "SNAPPY", "Encodings": "PLAIN(DICT_PAGE) RLE_DICTIONARY", "UncompressedSize": "100", "CompressedSize": "104" },
          {"Id": "1", "Values": "3", "StatsSet": "True", "Stats": {"NumNulls": "0", "Max": "c", "Min": "a" },
           "Compression": "SNAPPY", "Encodings": "PLAIN(DICT_PAGE) RLE_DICTIONARY", "UncompressedSize": "66", "CompressedSize": "70" }
        ]
     }
  ]
}
```

**Comment by wgtmac** [comment]
Should we break a new line for each item in `SortColumns`? It currently looks a little bit lengthy to me.

**Comment by mapleFU** [comment]
Update the output:

```json
{
  "FileName": "sort_columns.parquet",
  "Version": "2.6",
  "CreatedBy": "parquet-cpp-arrow version 16.1.0",
  "TotalRows": "6",
  "NumberOfRowGroups": "2",
  "NumberOfRealColumns": "2",
  "NumberOfColumns": "2",
  "Columns": [
     { "Id": "0", "Name": "a", "PhysicalType": "INT64", "ConvertedType": "NONE", "LogicalType": {"Type": "None"} },
     { "Id": "1", "Name": "b", "PhysicalType": "BYTE_ARRAY", "ConvertedType": "UTF8", "LogicalType": {"Type": "String"} }
  ],
  "RowGroups": [
     {
       "Id": "0",  "TotalBytes": "166",  "TotalCompressedBytes": "174",  "SortColumns": [
         {"column_idx":0, "descending":1, "nulls_first": 1},
         {"column_idx":1, "descending":0, "nulls_first": 0}
       ],  "Rows": "3",
       "ColumnChunks": [
          {"Id": "0", "Values": "3", "StatsSet": "True", "Stats": {"NumNulls": "1", "Max": "2", "Min": "1" },
           "Compression": "SNAPPY", "Encodings": "PLAIN(DICT_PAGE) RLE_DICTIONARY", "UncompressedSize": "100", "CompressedSize": "104" },
          {"Id": "1", "Values": "3", "StatsSet": "True", "Stats": {"NumNulls": "0", "Max": "c", "Min": "a" },
           "Compression": "SNAPPY", "Encodings": "PLAIN(DICT_PAGE) RLE_DICTIONARY", "UncompressedSize": "66", "CompressedSize": "70" }
        ]
     },
     {
       "Id": "1",  "TotalBytes": "166",  "TotalCompressedBytes": "174",  "SortColumns": [
         {"column_idx":0, "descending":1, "nulls_first": 1},
         {"column_idx":1, "descending":0, "nulls_first": 0}
       ],  "Rows": "3",
       "ColumnChunks": [
          {"Id": "0", "Values": "3", "StatsSet": "True", "Stats": {"NumNulls": "1", "Max": "2", "Min": "1" },
           "Compression": "SNAPPY", "Encodings": "PLAIN(DICT_PAGE) RLE_DICTIONARY", "UncompressedSize": "100", "CompressedSize": "104" },
          {"Id": "1", "Values": "3", "StatsSet": "True", "Stats": {"NumNulls": "0", "Max": "c", "Min": "a" },
           "Compression": "SNAPPY", "Encodings": "PLAIN(DICT_PAGE) RLE_DICTIONARY", "UncompressedSize": "66", "CompressedSize": "70" }
        ]
     }
  ]
}
```

**Comment by mapleFU** [comment]
@wgtmac @pitrou would you mind take a look?

**Review by wgtmac — APPROVED** [review]
+1

Thanks @mapleFU!

File: cpp/src/parquet/printer.cc [hunk 1]
```diff
@@ -142,6 +142,15 @@ void ParquetFilePrinter::DebugPrint(std::ostream& stream, std::list<int> selecte
     stream << "--- Total Bytes: " << group_metadata->total_byte_size() << " ---\n";
     stream << "--- Total Compressed Bytes: " << group_metadata->total_compressed_size()
            << " ---\n";
+    auto sorting_columns = group_metadata->sorting_columns();
```
Comment (pitrou) [inline cpp/src/parquet/printer.cc:145]: Nit
```suggestion
    const auto& sorting_columns = group_metadata->sorting_columns();
```

File: cpp/src/parquet/printer.cc [hunk 2]
```diff
@@ -285,6 +294,21 @@ void ParquetFilePrinter::JSONPrint(std::ostream& stream, std::list<int> selected
     stream << " \"TotalBytes\": \"" << group_metadata->total_byte_size() << "\", ";
     stream << " \"TotalCompressedBytes\": \"" << group_metadata->total_compressed_size()
            << "\", ";
+    auto row_group_sorting_columns = group_metadata->sorting_columns();
```
Comment (pitrou) [inline cpp/src/parquet/printer.cc:297]: Similar nit
```suggestion
    const auto& row_group_sorting_columns = group_metadata->sorting_columns();
```

**Review by pitrou — APPROVED** [review]
LGTM, just 2 nits

File: cpp/src/parquet/printer.cc (hunk 1 above)
Comment (mapleFU) [reply cpp/src/parquet/printer.cc:145]:
```c++
std::vector<SortingColumn> RowGroupMetaData::sorting_columns() const
```

Seems that this returns a vector directly, and just auto is ok here...

File: cpp/src/parquet/printer.cc (hunk 2 above)
Comment (mapleFU) [reply cpp/src/parquet/printer.cc:297]: ditto

**Pull request merged**